test/virtual/host.cpp

```diff
@@ -26,16 +156,29 @@ extern "C"
     {
         oe_result_t result = OE_FAILURE;
         oe_enclave_t* previous_enclave = _enclave;
+        uint64_t function_id = OE_ECALL_ID_NULL;
         _enclave = enclave;
         enclave->_set_enclave(enclave);
 
         if (!_enclave->is_outside_enclave(input_buffer, input_buffer_size) ||
-            !_enclave->is_outside_enclave(output_buffer, output_buffer_size))
+            !_enclave->is_outside_enclave(output_buffer, output_buffer_size) ||
+            !global_id)
         {
             result = OE_INVALID_PARAMETER;
             goto done;
         }
 
+        /*
+         * Look up the function id from the per-enclave table based on the
+         * global id. The global id is defined as a static variable in the
+         * oeedger8r-generated code. The function initializes the global id in
+         * the first invocation and uses the cached global id for the subsequent
+         * invocations.
+         */
+        if (oe_get_enclave_function_id(
```

Comment:
This needs careful acquire-release semantics.
oe_get_enclave_function_id could use double-checked locking with correct acquire-release semantics.